waiter/src/waiter/service_description.clj

```diff
@@ -635,14 +636,46 @@
   [service-description username]
   (assoc service-description "run-as-user" username "permitted-user" username))
 
-(defn service-token-references-stale?
-  "Returns true if every token used to access a service has been updated."
-  [token->token-hash source-tokens]
-  (and (seq source-tokens)
-       ;; safe assumption mark a service stale when every token used to access it is stale
-       (every? (fn [{:keys [token version]}]
-                 (not= (token->token-hash token) version))
-               source-tokens)))
+(let [hash-prefix "E-"]
+  (defn token-data->token-hash
+    "Converts the merged map of service-description and token-metadata to a hash."
+    [token-data]
+    (when (and (seq token-data)
+               (not (get token-data "deleted")))
+      (str hash-prefix (-> token-data
+                         (select-keys token-data-keys)
+                         (dissoc "previous")
+                         parameters->id)))))
+
+(defn retrieve-token-update-epoch-time
+  "Returns the time at which the specific version of a token went stale.
+   If no such version can be found, then the stale time is unknown and we, conservatively,
+   use the last known update time.
+   Else if the token is not stale or no data is known about the token, nil is returned."
+  [current-token-data token-version]
+  (let [token-data->update-time #(get % "last-update-time")
+        token-data->previous #(get % "previous")]
+    (utils/retrieve-update-time
+      token-data->token-hash token-data->update-time token-data->previous current-token-data token-version)))
+
+(defn retrieve-token-stale-info
+  "The provided source-tokens are stale if every token used to access a service has been updated.
+   Returns a map containing two keys:
+   - :stale? is true if the provided source-tokens are stale;
+   - :update-epoch-time is either nil or the most recent time any of the source tokens of
```

Comment:
Yes, it should read `earliest` instead of `recent`, my direction of reference was from when the tokens became stale. Sorry for the confusion, will correct.

Reply:
Further rephrased the comment